ghc-toolkit/boot-libs/rts/PrimOps.cmm

```diff
@@ -1441,12 +1441,82 @@ stg_takeMVarzh ( P_ mvar /* :: MVar a */ )
      * and wait until we're woken up.
      */
     if (StgMVar_value(mvar) == stg_END_TSO_QUEUE_closure) {
-        jump stg_raisezh(base_GHCziIOziException_blockedIndefinitelyOnMVar_closure);
+        //if (info == stg_MVAR_CLEAN_info) {
+        //    ccall dirty_MVAR(BaseReg "ptr", mvar "ptr");
+        //}
+
+        // We want to put the heap check down here in the slow path,
+        // but be careful to unlock the closure before returning to
+        // the RTS if the check fails.
+        ALLOC_PRIM_WITH_CUSTOM_FAILURE
+            (SIZEOF_StgMVarTSOQueue,
+             GC_PRIM_P(stg_takeMVarzh, mvar));
+
+        q = Hp - SIZEOF_StgMVarTSOQueue + WDS(1);
+
+        SET_HDR(q, stg_MVAR_TSO_QUEUE_info, CCS_SYSTEM);
+        StgMVarTSOQueue_link(q) = END_TSO_QUEUE;
+        StgMVarTSOQueue_tso(q)  = CurrentTSO;
+
+        if (StgMVar_head(mvar) == stg_END_TSO_QUEUE_closure) {
+            StgMVar_head(mvar) = q;
+        } else {
+            StgMVarTSOQueue_link(StgMVar_tail(mvar)) = q;
+            //ccall recordClosureMutated(MyCapability() "ptr",
```

Comment:
Same question above; we can have a dummy `recordClosureMutated` builtin function so even less hacks here.